### ISSUE TITLE
spec: Whitelists vs blacklists

### ISSUE THREAD

**Issue by schancel** [appc/spec#323]
I like whitelists from an idealogical viewpoint.  However, it seems to me that this potentially makes building images more difficult.  I wanted to open this up for some discourse.   Particularly, the issues that I have concerns around is snapshotting containers.  Seems like you're going to end up with an enumeration of every file in the dependencies, or have to implement some algorithms to prune that list.

**Comment by philips** [comment]
I am fine with having both whitelist and blacklist and saying you can only use one in an image.

**Comment by schancel** [comment]
Okay great.  I will submit a PR with a change to the spec as well as updates to the associated tools.

**Comment by philips** [comment]
@schancel Great! Thanks. I think @sgotti implemented some of the renderer stuff so FYI on this one.

**Comment by thockin** [comment]
+1 I had the same comment

**Comment by sgotti** [comment]
+1. Let me know if I can help you with the acirenderer code!

**Comment by schancel** [comment]
@sgotti If you're interested in doing the coding, feel free :)  I haven't gotten to any of it yet.  I am preoccupied with a few other things at the moment.  Does the SPEC.md change in the docs look okay to you?

**Comment by jonboulle** [comment]
@schancel " Does the SPEC.md change in the docs " what are you referring to here?

**Comment by derekchiang** [comment]
@sgotti @schancel is any of you still working on this?  I'm working on [a project](https://github.com/appc/acbuild) that is pretty much blocked by this issue, so I'd be happy to provide an implementation.

**Comment by jonboulle** [comment]
@derekchiang I think you should just go for it

**Comment by schancel** [comment]
@derekchiang Yes.   Sorry, this is not a blocker for me -- it is just something I thought needed to get done.   I had started to update the validators and whatnot, but didn't get very far.   Feel free to just nab the changes to the spec I made put them under your work.

https://github.com/appc/spec/commit/07e05458f0d9911b16be102723f183cbe979dce6

**Comment by sgotti** [comment]
@derekchiang I stopped waiting on https://github.com/appc/spec/issues/323#issuecomment-98961124 . I'll happily work or help on it. These are  busy weeks so I'll have more time on the next one. If it's an acbuild blocker feel free to take it. Thanks!

**Comment by derekchiang** [comment]
@schancel I think what we are proposing here is to have both a blacklist and a whitelist, but you can only use one of them in a given manifest, no?  Your commit seems to simply replace the whitelist with a blacklist.

**Comment by schancel** [comment]
Yes, I replaced it.  I don't see the value of a whitelist from a purely technical perspective.  (What are the use cases except hand-made ACIs?)  I am not opposed to having one optionally be in the spec though.

**Comment by derekchiang** [comment]
I have started putting together an implementation, but I just discovered one complication.  So according to the spec, the white list is supposed to be a "complete and absolute" set of paths.  That is, if `/a` is in the whitelist, it doesn't automatically mean that `/a/b` will be persisted.  You need to explicitly add `/a/b` to the whitelist to persist it.

With the blacklist however, if `/a` is in the blacklist, it doesn't seem to make sense to keep `/a/b` or any other files/directories under `/a`.  So there is an asymmetry here: while the whitelist doesn't affect the children of a directory, the blacklist does.

Does this sound OK to you guys?

**Comment by sgotti** [comment]
@derekchiang sorry for the delay. Seems reasonable to me that if a blacklist entry will automatically blacklist all of its children (it should just be clearly explained in the spec).

**Comment by jonboulle** [comment]
+1

**Comment by vbatts** [comment]
+1

**Comment by jonboulle** [comment]
@vbatts c'mon, don't be that guy, put up a PR
